language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newparam(:instance_initiated_shutdown_behavior) do
+    desc 'Whether the instance stops or terminates when you initiate shutdown from the instance.'
+    newvalues(:stop, :terminate)
+  end
+end
+```
+
+The singular form `newvalue` defines one possible value and takes a block. At catalog application time, if puppet determines the property is not insync, it will call the block to "sync" that resource's property. It's common to define multiple `newvalue` blocks, where each value's block calls an appropriate provider method.
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:ensure) do
+    newvalue(:running) do
+      provider.create unless provider.running?
+    end
+    newvalue(:stopped) do
+      provider.stop unless provider.stopped?
+    end
+  end
+end
+```
+
+The `newvalue` and `newvalues` methods can also be passed a regex. Puppet will compare explicit symbols/strings first, and if there are no matches, compare regex's. For example, the `package` type defines:
+
+```ruby
+Puppet::Type.newtype(:package) do
+  newproperty(:ensure) do
+    newvalue(:present)
+    newvalue(:absent)
+    ...
+    newvalue(/./)
+  end
+end
+```
+
+The last regex is used to match version strings, e.g `ensure => '1.2.3'`
+
+### aliasvalue
+
+Aliases a value to be the same as an existing value. For example, the package type aliases `installed` to be the same as `present`, because it's more natural to declare that a package is `installed`:
+
+```
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:ensure) do
+    newvalue(:present)
+    aliasvalue(:installed, :present)
+  end
+end
+```
+
+### defaultto
+
+Defines the default value for a parameter. If a value is specified, then `defaultto` creates a `default` method with a block that always returns the specified value.
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newparam(:instance_initiated_shutdown_behavior) do
+    newvalues(:stop, :terminate)
+    defaultto :stop
+  end
+end
+```
+
+If a block is specified, then the block is called to retrieve the default value at catalog application time. For example, the `filebucket` type defines a `server` parameter: 
+
+```ruby
+Puppet::Type.newtype(:filebucket) do
+  newparam(:server) do
+    desc "The server providing the remote filebucket service."
+    defaultto { Puppet[:server] }
+  end
+end
+```
+
+The default value of `Puppet[:server]` works for all agents, because the default value is resolved at catalog application time, not compilation time.
+
+### is_to_s/should_to_s/change_to_s
+
+Overrides log messages for a property at catalog application time. Common examples are convert an id to a human readable name, sorting multi-valued attributes, redacting passwords, etc.
+
+```ruby
+Puppet::Type.newtype(:user) do
+  newproperty(:password) do
+    def is_to_s(currentvalue)
+      return '[old password hash redacted]'
+    end
+  end
+end
+```
+
```

Comment:
I thought about not even mentioning how `ensurable` does that, because I wanted to focus on the `Puppet::Type` side of things. I can imagine a separate **provider** doc that talks about how retrieve/insync/sync/mk_resource_methods/etc work.